### PR TITLE
Update Constant/TextureBuffer implementation.

### PR DESCRIPTION
**Pull request by python3kgae**

### CODE REVIEW

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp [hunk 1]
```diff
@@ -2503,7 +2506,7 @@ static DxilResourceBase::Class KeywordToClass(const std::string &keyword) {
     return DxilResourceBase::Class::CBuffer;
 
   if (keyword == "TextureBuffer")
-    return DxilResourceBase::Class::SRV;
+    return DxilResourceBase::Class::CBuffer;
```
Comment (tex3d) [inline tools/clang/lib/CodeGen/CGHLSLMS.cpp:2509]: Will this be fixed up to SRV later?

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 2]
```diff
@@ -7736,6 +7775,14 @@ ExprResult HLSLExternalSource::LookupArrayMemberExprForHLSL(
 
 ExprResult HLSLExternalSource::MaybeConvertScalarToVector(_In_ clang::Expr* E) {
```
Comment (tex3d) [inline tools/clang/lib/Sema/SemaHLSL.cpp:7776]: TODO: rename function to reflect expanded functionality.

**Review by tex3d — APPROVED** [review]
Looks great!  One question re: Class for TextureBuffer, and a minor TODO (possibly post-commit).

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp (hunk 1 above)
Comment (python3kgae) [reply tools/clang/lib/CodeGen/CGHLSLMS.cpp:2509]: Yes.

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp [hunk 3]
```diff
@@ -3219,64 +3211,110 @@ void CGMSHLSLRuntime::AddConstant(VarDecl *constDecl, HLCBuffer &CB) {
...
+      Ty = QualType(Ty->getArrayElementTypeNoTypeQual(),0);
+      CB->SetRangeSize(UINT_MAX);
     }
+    CB->SetIsArray();
   }
 
+  QualType ResultTy = hlsl::GetHLSLResourceResultType(Ty);
+
+    // Search defined structure for resource objects and fail
+  if (CB->GetRangeSize() > 1 && IsResourceInType(CGM.getContext(), ResultTy)) {
```
Comment (tex3d) [inline tools/clang/lib/CodeGen/CGHLSLMS.cpp:3303]: Will this miss CBV[1] case? I think it should be checking the new bIsView instead.

**Comment by ehsannas** [comment]
I have just started working on updating the SPIR-V backend to work with these changes.

File: tools/clang/lib/AST/ASTContextHLSL.cpp [hunk 4]
```diff
@@ -831,6 +831,27 @@ CXXRecordDecl* hlsl::DeclareUIntTemplatedTypeWithHandle(
   return typeDeclBuilder.completeDefinition();
 }
 
+clang::CXXRecordDecl *
+hlsl::DeclareConstantBufferViewType(clang::ASTContext &context, bool bTBuf) {
+  // Create ConstantBufferView template declaration in translation unit scope.
+  // template<typename T> ConstantBuffer : public T {}
```
Comment (ehsannas) [inline tools/clang/lib/AST/ASTContextHLSL.cpp:837]: Stale comment?

File: tools/clang/lib/AST/ASTContextHLSL.cpp [hunk 5]
```diff
@@ -831,6 +831,27 @@ CXXRecordDecl* hlsl::DeclareUIntTemplatedTypeWithHandle(
...
+  DeclContext *DC = context.getTranslationUnitDecl();
+
+  BuiltinTypeDeclBuilder typeDeclBuilder(DC, bTBuf ? "TextureBuffer"
+                                                   : "ConstantBuffer");
+  TemplateTypeParmDecl *elementTemplateParamDecl =
+      typeDeclBuilder.addTypeTemplateParam("T");
```
Comment (ehsannas) [inline tools/clang/lib/AST/ASTContextHLSL.cpp:843]: `elementTemplateParamDecl` is an unused variable.

**Comment by ehsannas** [comment]
@python3kgae  The SPIR-V backend changes are done now.

@jaebaek Please review the SPIR-V backend commit in this PR (28773c71d463e1665285002e5234f5cdfafadef4). FYI, these were the tests that were failing before my commit: [log](https://ci.appveyor.com/project/antiagainst/directxshadercompiler/builds/35286048/job/h096l4h2arsp4rt5#L1823)

**Pull request merged**